OpenRA.Mods.RA2/Traits/AttackGarrisonedRV.cs

```diff
@@ -0,0 +1,222 @@
+		public readonly WAngle[] PortYaws = null;
+
+		[FieldLoader.Require]
+		[Desc("Fire port yaw cone angle.")]
+		public readonly WAngle[] PortCones = null;
+
+		public FirePort[] Ports { get; private set; }
+
+		[PaletteReference] public readonly string MuzzlePalette = "effect";
+
+		public override object Create(ActorInitializer init) { return new AttackGarrisonedRV(init.Self, this); }
+		public override void RulesetLoaded(Ruleset rules, ActorInfo ai)
+		{
+			if (PortOffsets.Length == 0)
+				throw new YamlException("PortOffsets must have at least one entry.");
+
+			if (PortYaws.Length != PortOffsets.Length)
+				throw new YamlException("PortYaws must define an angle for each port.");
+
+			if (PortCones.Length != PortOffsets.Length)
+				throw new YamlException("PortCones must define an angle for each port.");
+
+			Ports = new FirePort[PortOffsets.Length];
+
+			for (var i = 0; i < PortOffsets.Length; i++)
+			{
+				Ports[i] = new FirePort
+				{
+					Offset = PortOffsets[i],
+					Yaw = PortYaws[i],
+					Cone = PortCones[i],
+				};
+			}
+
+			base.RulesetLoaded(rules, ai);
+		}
+	}
+
+	public class AttackGarrisonedRV : AttackFollow, INotifyGarrisonerEntered, INotifyGarrisonerExited, IRender
+	{
+		public readonly new AttackGarrisonedRVInfo Info;
+		Lazy<BodyOrientation> coords;
+		List<Armament> armaments;
+		List<AnimationWithOffset> muzzles;
+		Dictionary<Actor, IFacing> paxFacing;
+		Dictionary<Actor, IPositionable> paxPos;
+		Dictionary<Actor, RenderSprites> paxRender;
+
+		public AttackGarrisonedRV(Actor self, AttackGarrisonedRVInfo info)
+			: base(self, info)
+		{
+			Info = info;
+			coords = Exts.Lazy(() => self.Trait<BodyOrientation>());
+			armaments = new List<Armament>();
+			muzzles = new List<AnimationWithOffset>();
+			paxFacing = new Dictionary<Actor, IFacing>();
+			paxPos = new Dictionary<Actor, IPositionable>();
+			paxRender = new Dictionary<Actor, RenderSprites>();
+		}
+
+		protected override Func<IEnumerable<Armament>> InitializeGetArmaments(Actor self)
+		{
+			return () => armaments;
+		}
+
+		void INotifyGarrisonerEntered.OnGarrisonerEntered(Actor self, Actor passenger)
+		{
+            // TODO - darky - Refactor - passenger -> garrisoner
+            paxFacing.Add(passenger, passenger.Trait<IFacing>());
+			paxPos.Add(passenger, passenger.Trait<IPositionable>());
+			paxRender.Add(passenger, passenger.Trait<RenderSprites>());
+			armaments.AddRange(
+				passenger.TraitsImplementing<Armament>()
+				.Where(a => Info.Armaments.Contains(a.Info.Name)));
+		}
+
+		void INotifyGarrisonerExited.OnGarrisonerExited(Actor self, Actor passenger)
+		{
+            // TODO - darky - Refactor - passenger -> garrisoner
+            paxFacing.Remove(passenger);
+			paxPos.Remove(passenger);
+			paxRender.Remove(passenger);
+			armaments.RemoveAll(a => a.Actor == passenger);
+		}
+
+		FirePort SelectFirePort(Actor self, WAngle targetYaw)
+		{
+			// Pick a random port that faces the target
```

Comment:
I'd wonder about this one. In RA2, garrison ports were associated with the passenger slots and ignored their own facing. OpenRA does not function this way and this opens the possibility to have more passengers/garrisonees using the same fireport the same time.

It was low priority to me to open a ticket about this behaviour difference in upstream but I wonder if we should change the behaviour here. I would opt doing so and then using the garrison code limited to allied actors upon the Battkle Fortress as well for better YR accuracy. @MustaphaTR, opinion?

Reply:
I'll be honest. I don't even know what a facing is.

Reply:
In that case I suggest to defer this change to a followup.